src/hotspot/cpu/riscv/macroAssembler_riscv.cpp

```diff
@@ -4246,6 +4246,64 @@ FCMP(double, d);
 
 #undef FCMP
 
+// Round double with mode
+
+void MacroAssembler::round_double_mode(FloatRegister dst, FloatRegister src, enum Round_double_mode round_mode, Register converted_dbl, Register mask, Register converted_dbl_masked)
+{
+
+  assert_different_registers(dst, src);
+  assert_different_registers(converted_dbl, mask, converted_dbl_masked);
+
+  // setting roundig mode to double->long (rm_direct) and long->double (rm_back) conversions
+  RoundingMode rm_direct, rm_back;
+  switch (round_mode) {
+    case Round_double_mode::rmode_ceil:
+      rm_direct = RoundingMode::rup;
+      rm_back = RoundingMode::rdn;
+      break;
+    case Round_double_mode::rmode_floor:
+      rm_direct = RoundingMode::rdn;
+      rm_back = RoundingMode::rup;
+      break;
+    case Round_double_mode::rmode_rint:
+      rm_direct = RoundingMode::rne;
+      rm_back = RoundingMode::rne;
+      break;
+    default:
+      ShouldNotReachHere();
+  }
+
+  // converted_dbl - is a register to store double converted to long int
+  // mask - is a register to create constant for comparsion
+  // converted_dbl_masked - is a register were we stroe modidfied result of double -> long int comparison
+  Label done, bad_val;
+
+  // generating constant (tmp2)
+  // tmp2 = 100...0000
+  addi(mask, zr, 1);
+  slli(mask, mask, 63);
+  // conversion from double to long
+  fcvt_l_d(converted_dbl, src, rm_direct);
```

Comment:
How about we use `fclass` [1] to check the special cases of input, then we can just do `fcvt.l.d` and `fcvt.d.l` for normal inputs. We can just check if the result of fclass hits NaN/infinity/+0/-0, if the input contains these values, we can just return the value without conversion.

1. https://github.com/riscv/riscv-isa-manual/blob/3a6edf7ebf6af9e6ad92ace865c0069090870c20/src/f-st-ext.adoc?plain=1#L487-L500

Reply:
Hi, thanks for your review. Also, we can use `fclass` to check cases NaN/+(-)INF/+(-)0.0 but we still need to check if value exeed `2^63 - 1 `(for positive input value) and `-2^63` (for negative one). So, we should leave check of converted value and we can add branch with a result of `fclass`. It will give an additional branch on regular values.